### PR TITLE
Fix goroutines stuck in write

### PR DESCRIPTION
**Pull request by martin-sucha**
This series of commits should fix stuck goroutines in case of network/server issues.

### CODE REVIEW

**Comment by martin-sucha** [comment]
When merging, do not squash as each commit has detailed description.

File: conn.go [hunk 1]
```diff
@@ -801,139 +803,210 @@ type callResp struct {
...
 	}
 
-	w.buffers = append(w.buffers, p)
-	for len(w.buffers) != 0 {
-		w.cond.Wait()
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case <-w.quit:
+		return 0, io.EOF // TODO: better error here?
```
Comment (matoous) [inline conn.go:919]: **nitpick**: forgotten TODO? TBH I have no idea what would be better 🤷

**Comment by matoous** [comment]
#683bf90 - LGTM
#f03d9e6 - will have to go through it a few more times 😅 
#cd071ad - LGTM
#a2a755d - LGTM but the test is still failing with:
```
==> Running cassandra scylla tests with args: -gocql.timeout=60s -proto=4 -rf=3 -clusterSize=2 -autowait=2000ms -compressor=snappy -gocql.cversion=3.11.4 -cluster=192.168.100.11,192.168.100.12
2021/04/09 17:33:29 logger.go:27: error: failed to connect to 192.168.100.11:9042 due to error: context canceled
2021/04/09 17:33:29 logger.go:27: error: failed to connect to 192.168.100.11:9042 due to error: Keyspace 'invalidKeyspace' does not exist
2021/04/09 17:33:29 logger.go:27: error: failed to connect to 192.168.100.12:9042 due to error: Keyspace 'invalidKeyspace' does not exist
2021/04/09 17:33:35 logger.go:27: error: failed to connect to 192.168.100.11:9042 due to error: context canceled
2021/04/09 17:33:40 logger.go:27: gocql: unable to dial control conn 127.0.0.1:1: dial tcp 127.0.0.1:1: connect: connection refused
2021/04/09 17:34:04 logger.go:27: error: failed to connect to 192.168.100.12:9042 due to error: context canceled
--- FAIL: TestTokenAwareConnPool (12.05s)
    cassandra_test.go:2099: Expected pool size 4 but was 3
2021/04/09 17:34:29 logger.go:27: error: failed to connect to 192.168.100.11:9042 due to error: context canceled
2021/04/09 17:34:45 common_test.go:63: error creating table table="CREATE TABLE gocql_test.errors_parse (id int primary key)" err=Cannot add already existing table "errors_parse" to keyspace "gocql_test"
2021/04/09 17:34:56 logger.go:27: reset ring after adding hosts: unsupported partitioner ''
2021/04/09 17:34:56 logger.go:27: reset ring after adding hosts: unsupported partitioner ''
--- FAIL: TestHostPolicy_TokenAwareLatency_NetworkStrategy_WithLatencyData (0.01s)
    tokenawarelatency_test.go:294: expected selected hosts to be ([05 07 04 06 03 08 01 00 02 09 10 11]) but was ([05 07 04 06 03 08 10 00 01 02 09 11]) instead
2021/04/09 17:34:56 logger.go:27: reset ring after adding hosts: unsupported partitioner ''
2021/04/09 17:34:56 logger.go:27: can't get metadata for keyspace "keyspace", skipped replica map update: some error
FAIL
exit status 1
FAIL	github.com/gocql/gocql	111.631s
Error: Process completed with exit code 1.
```

File: conn.go (hunk 1 above)
Comment (martin-sucha) [reply conn.go:919]: It was in the original code, so I kept it there. Another option is `ErrClosedConnection`.

**Comment by martin-sucha** [comment]
I can't reproduce the integration test failure locally when running with gocql_debug :disappointed: Will try again tomorrow.

**Pull request merged**